File: samples/AdventureWorksLT.Data.SqlServer/AdventureWorksLT.Data/ProductDescription.cs
using DevZest.Data;
using DevZest.Data.SqlServer;
using DevZest.Data.Annotations;

namespace DevZest.Samples.AdventureWorksLT
{
    public class ProductDescription : BaseModel<ProductDescription.PK>
    {
        [DbPrimaryKey("PK_ProductDescription_ProductDescriptionID", Description = "Primary key (clustered) constraint")]
        public sealed class PK : CandidateKey
        {
            public PK(_Int32 productDescriptionID)
                : base(productDescriptionID)
            {
            }
        }

        public class Key : Key<PK>
        {
            static Key()
            {
                Register((Key _) => _.ProductDescriptionID, _ProductDescriptionID);
            }

            protected sealed override PK CreatePrimaryKey()
            {
                return new PK(ProductDescriptionID);
            }

            public _Int32 ProductDescriptionID { get; private set; }
        }

        public class Ref : Ref<PK>
        {
            static Ref()
            {
                Register((Ref _) => _.ProductDescriptionID, _ProductDescriptionID);
            }

            public _Int32 ProductDescriptionID { get; private set; }

            protected override PK CreateForeignKey()
            {
                return new PK(ProductDescriptionID);
            }
        }

        public class Lookup : Projection
        {
            static Lookup()
            {
                Register((Lookup _) => _.Description, _Description);
            }

            public _String Description { get; private set; }
        }

        public static readonly Mounter<_Int32> _ProductDescriptionID = RegisterColumn((ProductDescription _) => _.ProductDescriptionID);
        public static readonly Mounter<_String> _Description = RegisterColumn((ProductDescription _) => _.Description);

        protected sealed override PK CreatePrimaryKey()
        {
            return new PK(ProductDescriptionID);
        }

        [Identity]
        [DbColumn(Description = "Primary key for ProductDescription records.")]
        public _Int32 ProductDescriptionID { get; private set; }

        [Required]
        [SqlNVarChar(400)]
        [DbColumn(Description = "Description of the product.")]
        public _String Description { get; private set; }
    }
}

File: samples.vb/AdventureWorksLT.SqlServer/AdventureWorksLT/ProductDescription.vb
Public Class ProductDescription
    Inherits BaseModel(Of ProductDescription.PK)

    <DbPrimaryKey("PK_ProductDescription_ProductDescriptionID", Description:="Primary key (clustered) constraint")>
    Public NotInheritable Class PK
        Inherits CandidateKey

        Public Sub New(productDescriptionID As _Int32)
            MyBase.New(productDescriptionID)
        End Sub
    End Class

    Public Class Key
        Inherits Key(Of PK)

        Shared Sub New()
            Register(Function(x As Key) x.ProductDescriptionID, _ProductDescriptionID)
        End Sub

        Protected Overrides Function CreatePrimaryKey() As PK
            Return New PK(ProductDescriptionID)
        End Function

        Private m_ProductDescriptionID As _Int32
        Public Property ProductDescriptionID As _Int32
            Get
                Return m_ProductDescriptionID
            End Get
            Private Set
                m_ProductDescriptionID = Value
            End Set
        End Property
    End Class

    Public Class Ref
        Inherits Ref(Of PK)

        Shared Sub New()
            Register(Function(x As Ref) x.ProductDescriptionID, _ProductDescriptionID)
        End Sub

        Private m_ProductDescriptionID As _Int32
        Public Property ProductDescriptionID As _Int32
            Get
                Return m_ProductDescriptionID
            End Get
            Private Set
                m_ProductDescriptionID = Value
            End Set
        End Property

        Protected Overrides Function CreateForeignKey() As PK
            Return New PK(ProductDescriptionID)
        End Function
    End Class

    Public Class Lookup
        Inherits Projection

        Shared Sub New()
            Register(Function(x As Lookup) x.Description, _Description)
        End Sub

        Private m_Description As _String
        Public Property Description As _String
            Get
                Return m_Description
            End Get
            Private Set
                m_Description = Value
            End Set
        End Property
    End Class

    Public Shared ReadOnly _ProductDescriptionID As Mounter(Of _Int32) = RegisterColumn(Function(x As ProductDescription) x.ProductDescriptionID)
    Public Shared ReadOnly _Description As Mounter(Of _String) = RegisterColumn(Function(x As ProductDescription) x.Description)

    Protected NotOverridable Overrides Function CreatePrimaryKey() As PK
        Return New PK(ProductDescriptionID)
    End Function

    Private m_ProductDescriptionID As _Int32
    <Identity>
    <DbColumn(Description:="Primary key for ProductDescription records.")>
    Public Property ProductDescriptionID As _Int32
        Get
            Return m_ProductDescriptionID
        End Get
        Private Set
            m_ProductDescriptionID = Value
        End Set
    End Property

    Private m_Description As _String
    <Required>
    <SqlNVarChar(400)>
    <DbColumn(Description:="Description of the product.")>
    Public Property Description As _String
        Get
            Return m_Description
        End Get
        Private Set
            m_Description = Value
        End Set
    End Property
End Class
